libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,476 @@
+    LOG_DEBUG("dt_type\n");
+  } else {
+    uint8_t *rx_data;
+    size_t len;
+    if (!prv_read_rx_buffer(&rx_data, &len)) {
+      event_raise_no_data(DATAGRAM_EVENT_CRC);
+    } else {
+      memcpy(&dt->crc, rx_data, CRC_SIZE_BYTES);
+      event_raise_no_data(DATAGRAM_EVENT_DT_TYPE);
+      LOG_DEBUG("dt_type\n");
+    }
+  }
+}
+
+static void prv_process_dt_type(Fsm *fsm, const Event *e, void *context) {
+  CanDatagram *dt = context;
+  if (s_store.mode == CAN_DATAGRAM_MODE_TX) {
+    memcpy(s_can_buffer, &dt->dt_type, DT_TYPE_SIZE_BYTES);
+    s_store.tx_cb(s_can_buffer, sizeof(dt->dt_type), false);
+    event_raise_no_data(DATAGRAM_EVENT_DST_LEN);
+  } else {
+    uint8_t *rx_data;
+    size_t len;
+    if (!prv_read_rx_buffer(&rx_data, &len)) {
+      event_raise_no_data(DATAGRAM_EVENT_DT_TYPE);
+      LOG_DEBUG("dst_len\n");
+    } else {
+      memcpy(&dt->dt_type, rx_data, DT_TYPE_SIZE_BYTES);
+      event_raise_no_data(DATAGRAM_EVENT_DST_LEN);
+      LOG_DEBUG("dst_len\n");
+    }
+  }
+}
+
+static void prv_process_dst_len(Fsm *fsm, const Event *e, void *context) {
+  CanDatagram *dt = context;
+  if (s_store.mode == CAN_DATAGRAM_MODE_TX) {
+    memcpy(s_can_buffer, &dt->destination_nodes_len, DEST_LEN_SIZE_BYTES);
+    s_store.tx_cb(s_can_buffer, sizeof(dt->destination_nodes_len), false);
+    event_raise_no_data(DATAGRAM_EVENT_DST);
+    LOG_DEBUG("dst\n");
+  } else {
+    uint8_t *rx_data;
+    size_t len;
+    if (!prv_read_rx_buffer(&rx_data, &len)) {
+      event_raise_no_data(DATAGRAM_EVENT_DST_LEN);
+    } else {
+      memcpy(&dt->destination_nodes_len, rx_data, DEST_LEN_SIZE_BYTES);
+      event_raise_no_data(DATAGRAM_EVENT_DST);
+      LOG_DEBUG("dst\n");
+    }
+  }
+}
+
+static void prv_process_dst(Fsm *fsm, const Event *e, void *context) {
+  CanDatagram *dt = context;
+  uint8_t dst_len = dt->destination_nodes_len;
+  if (s_store.mode == CAN_DATAGRAM_MODE_TX) {
+    uint8_t dst_bytes_to_send = 0;
+    // TX max of 4 messages at a time
+    for (int msg = 0; msg < MAX_CAN_TX; msg++) {
+      dst_bytes_to_send = (dst_len - s_store.tx_bytes_sent < CAN_BUFFER_SIZE)
+                              ? (dst_len - s_store.tx_bytes_sent)
+                              : CAN_BUFFER_SIZE;
+      memcpy(s_can_buffer, dt->destination_nodes + s_store.tx_bytes_sent,
+             (size_t)dst_bytes_to_send);
+      s_store.tx_bytes_sent += dst_bytes_to_send;
+      s_store.tx_cb(s_can_buffer, dst_bytes_to_send, false);
+      if (s_store.tx_bytes_sent == dst_len) {
+        event_raise_no_data(DATAGRAM_EVENT_DATA_LEN);
+        LOG_DEBUG("data len\n");
+        s_store.tx_bytes_sent = 0;
+        return;
+      }
+    }
+    event_raise_no_data(DATAGRAM_EVENT_DST);
+  } else {
+    uint8_t *rx_data;
+    size_t len;
+    if (!prv_read_rx_buffer(&rx_data, &len)) {
+      event_raise_no_data(DATAGRAM_EVENT_DST);
+    } else {
+      memcpy(dt->destination_nodes + s_store.rx_bytes_read, rx_data, len);
```

Comment:
I think we should take in a node id and then return early from rxing any datagrams that don't have a destination node matching our node id since they aren't meant for us. That also might mean we don't have to store the destination nodes!